### ISSUE TITLE
tcomb lists + es6 classes

### ISSUE THREAD

**Issue by timoxley** [gcanti/tcomb#92]
tcomb lists don't currently play nice with es6 classes due to inability to invoke classes without `new`:

```
$ babel-node
> const t = require('tcomb')
> class Person extends t.struct({name: t.Str}) {}
> const a = new Person({name: 'alice'})
> const b = new Person({name: 3})
/Users/timoxley/Projects/tests/tcomb-classes/node_modules/tcomb/index.js:15
    throw new TypeError(message);
          ^
TypeError: Invalid argument `value` = `3` supplied to irreducible type `Str`
    at Object.fail (/Users/timoxley/Projects/tests/tcomb-classes/node_modules/tcomb/index.js:15:11)
    at assert (/Users/timoxley/Projects/tests/tcomb-classes/node_modules/tcomb/index.js:21:15)
    at Irreducible (/Users/timoxley/Projects/tests/tcomb-classes/node_modules/tcomb/index.js:174:7)
    at Person.Struct (/Users/timoxley/Projects/tests/tcomb-classes/node_modules/tcomb/index.js:260:21)
    at new Person (repl:12:17)
    at repl:3:9
    at Object.exports.runInThisContext (vm.js:74:17)
    at _eval (/usr/local/lib/node_modules/babel/bin/_babel-node:54:13)
    at REPLServer.replEval (/usr/local/lib/node_modules/babel/bin/_babel-node:136:14)
    at bound (domain.js:254:14)
> const People = t.list(Person)
> const c = new People([{name: 'cathy'}])
repl:3
nce, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeEr
                                                                    ^
TypeError: Cannot call a class as a function
    at _classCallCheck (repl:3:106)
    at Person (repl:9:5)
    at new List (/Users/timoxley/Projects/tests/tcomb-classes/node_modules/tcomb/index.js:513:18)
    at repl:3:9
    at Object.exports.runInThisContext (vm.js:74:17)
    at _eval (/usr/local/lib/node_modules/babel/bin/_babel-node:54:13)
    at REPLServer.replEval (/usr/local/lib/node_modules/babel/bin/_babel-node:136:14)
    at bound (domain.js:254:14)
    at REPLServer.runBound [as eval] (domain.js:267:12)
    at REPLServer.<anonymous> (repl.js:279:12)
>
```

Code in question:

```js
      for (var i = 0, len = value.length ; i < len ; i++ ) {
        var actual = value[i];
        arr.push(type(actual, mut)); // <– calling `type` without `new`
      }
```

Not sure good solution here though since AFAIK it's impossible to detect whether something should be invoked with `new` or not.

**Comment by timoxley** [comment]
Actually, thinking on this any code passed as the Type that isn't a constructor is unlikely to have specific requirements about `thisArg` anyway so it should be safe to just invoke `new type(actual, mut)`

**Comment by gcanti** [comment]
AFAIK invoking `new type(actual, mut)` returns always an object:

```js

// the Str irreducible
var Str = function (value) {
  // assert(value is a string...)
  return value;
};

console.log(Str('a')); // => 'a'
console.log(new Str('a')); // => {}, should be 'a'
```

**Comment by timoxley** [comment]
@gcanti oh, you're right, I forgot about that case, of course. ugh, then I don't know of any a way around it.

**Comment by timoxley** [comment]
This is ugly but I see no other way around it… could you please special case instantiation of classes using something like https://github.com/miguelmota/is-class

I note you have zero dependencies so I'm not sure how you want to go about including that though.

**Comment by gcanti** [comment]
AFAIK only structs require `new`, what about checking the kind using the `meta` infos?

```js
function create(type, value, mut) {
  return type.meta.kind === 'struct' ? new type(value, mut) : type(value, mut);
}

...

for (var i = 0, len = value.length ; i < len ; i++ ) {
  var actual = value[i];
  arr.push(create(type, actual, mut));
}
```

**Comment by gcanti** [comment]
I made a quick test adding the `create` function, seems to work

```js
import t from 'tcomb';

class Person extends t.struct({name: t.Str}) {}

console.log(Person.meta); // => { kind: 'struct', props: { name: { [Function: Irreducible] meta: [Object], displayName: 'Str', is: [Function] } },  name: '{name: Str}' }

const a = new Person({name: 'alice'});
const People = t.list(Person);
const c = People([{name: 'cathy'}]); // => ok
```

**Comment by timoxley** [comment]
@gcanti lgtm.

**Comment by tuchk4** [comment]
@gcanti Yo
In suggested way I have issue with updating models:

```js
var SitemapStruct = new tcomb.struct({
  id: tcomb.Num,
  name: tcomb.Str
}, 'sitemap');

class SitemapModel extends SitemapStruct {
  doSomething() {

  }
}

var model = new SitemapModel({
  id: 1,
  name: 'test'
});

// model.doSomething == function

var updatedModel = SitemapStruct.update(model, {
  id: {
    $set: 5
  }
});

// updatedModel.doSomething != function
```

**Comment by tuchk4** [comment]
And it will be good to implement more friendly classes support. Maybe something like:

```js
class Sitemap extends tcomb.struct {
  constructor(data) {
    super(data, {
      id: tcomb.Num,
      name: tcomb.Str
    });
  }

  doSomething() {

  }
}
```

And prototype methods should also work with **tcomb.list**. I have troubles with it because this is the array.

**Comment by gcanti** [comment]
Well, AFAIK it's how `extends` works, it copies static methods. 
So `SitemapStruct.update === SitemapStruct.update` and `SitemapStruct.update` creates a `SitemapStruct` instance.
I never used `class` and `extends` with tcomb, maybe I'm missing some useful pattern: is there a particular reason to do

```js
var SitemapStruct = new tcomb.struct({
  id: tcomb.Num,
  name: tcomb.Str
}, 'sitemap');

class SitemapModel extends SitemapStruct {
  doSomething() {

  }
}
```

instead of

```js
var SitemapStruct = new tcomb.struct({
  id: tcomb.Num,
  name: tcomb.Str
}, 'sitemap');

SitemapStruct.prototype.doSomething = function () {
  
};
```

?

**Comment by tuchk4** [comment]
@gcanti Yes example with prototype will work. 

But it will be cool to develop in es6 way

**Comment by tuchk4** [comment]
When you develop application using ES6 - and the in some modules you meet *prototype* - this is like step into the past :)

Also base model classes could be very helpful. For example:

```js

import setter from 'lodash/object/set';

class BaseModel extends tcomb.struct {
  set(key, value) {
    var config = setter({}, key, {
      $set: value
    });

    return this.constructor.update(this, config);
  }

  push(key, value) {
    var config = setter({}, key, {
      $push: value
    });

    return this.constructor.update(this, config);
  }
}

class Sitemap extends BaseModel {
  static stuct() {
    return {
      id: tcomb.Num,
      name: tcomb.Str,
      nodes: tcomb.list(Sitemap)
    }
  }

  addChild(id, name) {
    return this.push('nodes', {id, name});
  }
}

var sitemap = new Sitemap({
  id: 1,
  name: 'test',
  nodes: [{
    id: 2,
    name: 'test 2'
  }, {
    id: 3,
    name: 'test 3'
  }]
});

sitemap = sitemap.addChild(4, 'test 4');
```

**Comment by dmaslov** [comment]
+1 for ES6 support

**Comment by tuchk4** [comment]
@gcanti any ideas?

**Comment by gcanti** [comment]
@tuchk4 Not sure I'm follow, what's your goal here? ES6 classes are nice syntactic sugar but your last snippet seems even less verbose if written in ES5

```js
var Sitemap = t.struct({
  id: t.Num,
  name: t.Str,
  nodes: t.Any
});

Sitemap.meta.props.nodes = t.list(Sitemap); // recursive type

Sitemap.prototype.addChild = function (id, name) {
  return Sitemap.update(this, {
    nodes: {
      $push: [{id, name, nodes: []}]
    }
  });
};

var sitemap = new Sitemap({
  id: 1,
  name: 'test',
  nodes: [{
    id: 2,
    name: 'test 2',
    nodes: []
  }, {
    id: 3,
    name: 'test 3',
    nodes: []
  }]
});

console.log(sitemap);
console.log(sitemap.addChild(4, 'test 4'));
```

**Comment by tuchk4** [comment]
@gcanti  Sure this will work. But the main goal - develop in ES6 way.

tcomb is great component and it should meet the modern approach.

**Comment by mvolodin** [comment]
@tuchk4 + for es6 support for tcom

**Comment by PavloKovalov** [comment]
Absolutely agree with @tuchk4 
ES6 support will be super useful

**Comment by WolfMessing** [comment]
+

**Comment by balkonsmoke** [comment]
+1 for @tuchk4 advice

@gcanti I believe that your example will look much better in ES6, e.g.
```javascript
class Sitemap extends t.struct {
  static props = {
    id:    t.Num,
    name:  t.Str,
    nodes: t.list(Sitemap)
  };

  addChild(id, name) {
    return Sitemap.update(this, {
      nodes: {
        $push: [{id, name, nodes: []}]
      }
    })
  }
}
```

**Comment by timoxley** [comment]
@gcanti did this actually get resolved?

**Comment by gcanti** [comment]
Hi @timoxley, what you mean? The issue was fixed by ba6e7d3

**Comment by javascriptpoet** [comment]
i wrote this function to create es6 type. Its not debugged yet but perhaps it might be useful or give ideas to somebody else. It allows to define es6 class in a very granular way including instance props, constructor props and proto methods for all extended classes.
`export default ({externals:{t}})=>({
    methods={},
    instanceProps={},
    constructorProps={},
    extendeeType,
    extendeeClass
})=>{
    const myType=Object.assign(
        t.refinement(t.Function,(C)=>{
            if(!C.constructor) return false;
            return checkMyMethods(C.prototype)
        },'class'),
        {methods, instanceProps, constructorProps, extendeeType,of,isMyInstance,checkMyMethods,
            extendeeClass:(!extendeeClass && !extendeeType)?Object.constructor:extendeeClass
        }
    );
    const isMyInstance=(c)=>{
        t.struct(instanceProps).is(c);
        if(extendeeType)extendeeType.isMyInstance(c)
    };
    const checkMyMethods=(proto)=>{
        if(!proto) return false;
        if(!t.struct(methods,{strict:true}).is(_.omit(proto,'constructor'))) return false;
        if(extendeeType) return extendeeType.checkMyMethods(Object.getPrototypeOf(proto));
        if(extendeeClass) return extendeeClass===proto.constructor
    };
    const of=(C)=>{
        myType.is(C);
        return class extends C {
            constructor(props){
                super(t.struct(constructorProps)(props));
                myType.isMyInstance(this)
            }
        }
    };
    return myType;
}`
